pystac/item_collection.py

```diff
@@ -0,0 +1,125 @@
+from copy import deepcopy
+from pystac.errors import STACTypeError
+from typing import Any, Dict, Iterator, List, Optional, Sized, Iterable
+
+import pystac
+from pystac.utils import make_absolute_href, is_absolute_href
+from pystac.serialization.identify import identify_stac_object_type
+
+
+class ItemCollection(Sized, Iterable[pystac.Item]):
+    """Implementation of a GeoJSON FeatureCollection whose features are all STAC
+    Items.
+
+    All :class:`~pystac.Item` instances passed to the :class:`~ItemCollection` instance
+    during instantiation are cloned and have their ``"root"`` URL cleared. Instances of
+    this class are iterable and sized (see examples below).
+
+    Any additional top-level fields in the FeatureCollection are retained in
+    :attr:`~ItemCollection.extra_fields` by the :meth:`~ItemCollection.from_dict` and
+    :meth:`~ItemCollection.from_file` methods and will be present in the serialized file
+    from :meth:`~ItemCollection.save_object`.
+
+    Examples:
+
+        Loop over all items in the ItemCollection
+
+        >>> item_collection: ItemCollection = ...
+        >>> for item in item_collection:
+        ...     ...
+
+        Get the number of Items in the ItemCollection
+
+        >>> length: int = len(item_collection)
+
+    """
+
+    items: List[pystac.Item]
+    """The list of :class:`pystac.Item` instances contained in this
+    ``ItemCollection``."""
+
+    extra_fields: Dict[str, Any]
+    """Dictionary containing additional top-level fields for the GeoJSON
+    FeatureCollection."""
+
+    def __init__(
+        self, items: List[pystac.Item], extra_fields: Optional[Dict[str, Any]] = None
+    ):
+        self.items = [item.clone() for item in items]
```

Comment:
I think this is the proper way to do this so that a developer doesn't create an ItemCollection without knowing it could modify the original Items.

However, in most cases it's unnecessary since the most common use case is to read an ItemCollection from somewhere else using the `from_dict` method where the Items were just created.

@tomaugspurger wrote up a related issue, and I fixed some of it in pystac-client by having an option to just fetch pages as JSON rather than always ItemCollection, but when an ItemCollection is created it's going to be hit by the deepcopy after creation issue here.

Reply:
https://github.com/stac-utils/pystac-client/issues/49

Reply:
Maybe it would make sense to have a `clone_items` parameter that defaults to `True` that would determine whether or not we clone the Items here. If `clone_items=False` then we would also want to make sure that we don't alter any links automatically, since that could cause problems in the associated Collections/Catalogs.

Reply:
Another idea I just had is to reconsider the whole initialization of an ItemCollection.

The __init__ class could take in JSON instead of expecting them to already be Items, and from_dict would pass in the features dictionary to __init__ as JSON and the Item() constructor occurs here in the ItemCollection constructor.

This way there wouldn't be a double hit and the user wouldn't have to do anything different. This would be the primary way of using the class.   

But then constructing an ItemCollection is harder...there could be a separate factory function but that would have to deserialize the Items then call  __init__.

How about we change Items to be `Union[List[Dict], List[Item]] and handle both cases?  from_dict won't do the Item() construction.

Reply:
I want to make sure I'm understanding the potential use-cases that we want to support here. Here's my summary of the use-cases for an ItemCollection, let me know if I'm missing anything:

1. Deserialize a `pystac.ItemCollection` directly from either a JSON FeatureCollection file or a JSON FeatureCollection-like `dict`

    As per the suggestion in stac-utils/pystac-client#49, we may want to add an option to `pystac.Item.from_dict` to skip the `deepcopy` and then also proxy that option to the `pystac.ItemCollection.from_dict` method to improve performance.

2. Create an `ItemCollection` from an iterable of `pystac.Item` instances.

    In this case, we want to provide an option to either clone the items (to protect against mutating the originals) or not clone them (for improved performance). I also realized as I was typing this that we should relax the typing on the `items` parameter from `List[pystac.Item]` to `Iterable[pystac.Item]` since the current code will handle both equally well.

> How about we change Items to be `Union[List[Dict], List[Item]] and handle both cases? from_dict won't do the Item() construction.

I think the only other use-case that this would support would be:

3. Creating an `ItemCollection` from a list of STAC Item-like dictionaries.

Do you think there's a need to be able to do this independent of the first use-case where we are just deserializing a full ItemCollection? If so, then it would definitely be worth changing the `__init__` signature, but if not then maybe we just keep it as-is for simplicity.

Reply:
Added a `clone_items` parameter to `ItemCollection.__init__` in 9b1fed25c879511e91ead274df5d5244520f34b8

Reply:
> I also realized as I was typing this that we should relax the typing on the `items` parameter from `List[pystac.Item]` to `Iterable[pystac.Item]` since the current code will handle both equally well.

Allowed `Iterable[pystac.Item]` for `items` argument instead of more strict `List[pystac.Item]` in 31beb9de97dd7370dab528f7d05b7d53f7003c69

Reply:
The first use case is certainly the most common - reading from a file or an http endpoint, and in these cases there's no need to deepcopy or clone anything. I think it will be far more rare to create an Item from a set of pystac.Item objects.

Theres another use case for ItemCollections, which is adding one or more together (e.g., pages of search results), maybe that's just an append to the features.

Looking at your recent changes with the clone_items argument, I think that should get passed on from .from_dict with kwargs, and "from_file" should automatically pass long `clone_items=False` to `.to_dict()`

Reply:
> I think it will be far more rare to create an Item from a set of pystac.Item objects.

Yeah, I agree we probably wouldn't use this directly very often, but I'm thinking we would use this when converting to/from `Catalog` and `Collection` instances. In that case we would probably want to be able to pass in an Iterable of `Item` instances and either clone them or not depending on the specific use-case. Something like:

```python
def manipulate_items(item: Item) -> Item:
    # Manipulate links and anything else we need here

item_collection = ItemCollection(items=map(manipulate_items, collection.get_items()))
```
Similarly, for adding `ItemCollection`s together we would already have `Item` instances, so we would probably want to just pass those Instances in directly without going to/from a dictionary.

Maybe the best course here is to change `__init__` to accept an Iterable of either Items or dicts, as you suggested, to cover our bases.

> Looking at your recent changes with the clone_items argument, I think that should get passed to **init** from `from_dict` (through kwargs), and `from_file` should automatically pass long `clone_items=False` to `.to_dict()`

Good catch, I'll update this.

Reply:
Can be initialized with an Iterable of either `Item` or `dict` instances as of 702a26f4a9a510d604dd223b490bb27a859ae898

Reply:
Ability to add `ItemCollection` instances in 1ead0d573c6f7356e947583aebec5c9c22a0987d.

This will combine unique `Items` into a new `ItemCollection`, so an `Item` that appears in 2 `ItemCollections` (determined by object equality) will only appear once in their sum (see tests or docstring for examples).

Reply:
> > Looking at your recent changes with the clone_items argument, I think that should get passed to **init** from `from_dict` (through kwargs), and `from_file` should automatically pass long `clone_items=False` to `.to_dict()`
> 
> Good catch, I'll update this.

I realized that there's no reason to clone Items in `from_dict` since they are being created within that method and therefore have no other references, so I just added `clone_items=False` to that method in 698c54bf94120f4896da8fb679b53a9f6c5dc76d

Reply:
> 1. As per the suggestion in [stac-utils/pystac-client#49](https://github.com/stac-utils/pystac-client/issues/49), we may want to add an option to `pystac.Item.from_dict` to skip the `deepcopy` and then also proxy that option to the `pystac.ItemCollection.from_dict` method to improve performance.

If it's okay with you, I may track this as a separate issue since I feel like the scope of this PR has grown enough already.

Reply:
@matthewhanson I think we have all of the use-cases that we discussed here in the current iteration. Let me know if you agree and I'll resolve this conversation.

Reply:
Based on our conversation today I changed `ItemCollection.__init__` to default to `clone_items=False`, which closes out the issues here. I'll resolve this conversation and re-request a review.